backends/aiger/xaiger.cc

```diff
@@ -355,7 +355,7 @@ struct XAigerWriter
 				}
 			}
 
-			for (auto port_name : r.first->second) {
+			for (const auto &port_name : r.first->second) {
```

Comment:
This is literally creating a pointer to an int instead of copying the int by value. That can't possibly be an improvement. There are a few more places in this patch where this is done.

Reply:
If it is a `const integer &`,  the compiler automatically choose to do the right thing choosing a copy or reference (when it doesn't have to worry about aliasing).

I see how this might not _look_ optimal, but there is no difference in what the compiler generates here ( Play with the reference `&` here: https://godbolt.org/z/v-Huvy ).

As someone reading the code, it is a readability improvement because they don't have to worry about if this is an potentially expensive copy or not, as the compiler will choose the fastest way with the const reference, but might not have a chance to do the fastest when it is a copy.

Reply:
The compilerexplorer code you have posted here is in no way representative for the code in question. The code in question contains external function calls interleaved with uses of `port_name`, thus the compiler has to re-fetch the value of `port_name` multiple times.

Please stop posting these misleading code snippets that seem to almost intentionally avoid the specific issue I've been pointing out numerous times now in the code in question.

Reply:
To be fair, your immediate raised concern in this example was that this was using a const reference to the int-wrapping `IdString`; so I responded directly to that and wanted to show in a reduced example that this is not typically a concern: the compiler will not necessarily dereference a const ref every time.

In generalization, when we can use a const reference, the compiler will have a chance to do this optimization, but if we force it to do a copy, the compiler might not have a chance to do that in case as it will have to call an expensive copy constructor. But you are also right that this requires enough visibility for the compiler to know that it is safe.

This very loop is actually a good example: in its current implementation, the `IdString` is not just a simple integer, there is also some global reference counting going on in the constructor/destructor (unless `YOSYS_NO_IDS_REFCNT` is defined (which it is not in the default compilation)). So in this loop, with the forced copy, there is actually one more instance of incrementing/decrementing the reference going on, making this loop slower with the copy vs. the const reference.

Since you're skeptical of my potentially biased godbolt-reductions, we can observe this directly in the actual piece of compiled code; do the following once with the stock xaiger.cc, then edit line 358 to use a const reference: `for (const auto &port_name : r.first->second)`

This command-line then would cut out the approximate area of code we're interested in:
```
make && objdump -l -C -d -S ./backends/aiger/xaiger.o | awk '/xaiger.cc:358/ { do_print=1; }  { if (do_print) print $0; } /xaiger.cc:388/ { do_print=0; }' > xaiger-loop.1
```
With a stock compilation with clang 8.0 this shows that there is at least one round more of reference counting going on. In that case the reference version clearly has an advantage.

This is an example where an unassumingly expensive constructor makes the difference that this pull request intends to address: avoiding these copies so that the compile has a chance to generate more optimal code.

If `YOSYS_NO_IDS_REFCNT` is defined and `IdString` is actually boiling down to an integer, the code is similar but requires a few more indirections (but I am not sure if in that case the clang-tidy would've suggested it).

Reply:
>  there is also some global reference counting going on

which, as you also point out, can be turned off using `YOSYS_NO_IDS_REFCNT`, which is what should be done if one is concerned about ref count performance. thus a change that improves performance for enabled reference counting, but worsens it with disabled reference counting, is counterproductive, because it worsens performance for those who care the most.